### ISSUE TITLE
为什么攻击和训练都只在CPU上运行

### ISSUE THREAD

**Issue by K0zero** [zhhiyuan/pretrain-model-attack#2]
我已经在config.py中注明了使用GPU
并且在开始打印user config中
打印出的是GPU

**Comment by K0zero** [comment]
但是只有CPU每次都跑满利用率
而GPU一直空闲

**Comment by zhhiyuan** [comment]
请先查看cuda是否可用

**Comment by K0zero** [comment]
> 请先查看cuda是否可用

我用torch.cuda.is_available() 显示True
并且torch.cuda.get_device_name(0) 显示出了 GPU型号
我还特地下载了预处理模型，但是单像素攻击（打印sucess rate和RGB）时候GPU利用率几乎为零

**Comment by zhhiyuan** [comment]
可以查看像素攻击代码，映像中其使用的是进化算法进行迭代攻击，无法跑在GPU上。

**Comment by K0zero** [comment]
> 可以查看像素攻击代码，映像中其使用的是进化算法进行迭代攻击，无法跑在GPU上。

那请问PGD攻击是可以在GPU上运行的吗
或者有没有办法改为使用GPU运行攻击

**Comment by zhhiyuan** [comment]
此仓库的PGD由于时间过久，忘记能否在GPU上运行，可以将下面一段代码替换attck/PGDattack中的内容，调用下面的代码可以在GPU上运行。

from torch import nn
import torch
import torch.nn.functional as F
class AttackPGD(nn.Module):
    def __init__(self, basic_net,epsilon=8.0/255,num_steps=10, step_size=2.0/255):
        super(AttackPGD, self).__init__()
        self.basic_net = basic_net
        self.step_size = step_size
        self.epsilon = epsilon
        self.num_steps = num_steps
    def forward(self, inputs, targets):
        x = inputs.detach()
        x = x + torch.zeros_like(x).uniform_(-self.epsilon, self.epsilon)
        for i in range(self.num_steps):
            x.requires_grad_()
            with torch.enable_grad():
                loss = F.cross_entropy(self.basic_net(x), targets, size_average=False)
            grad = torch.autograd.grad(loss, [x])[0]
            x = x.detach() + self.step_size*torch.sign(grad.detach())
            x = torch.min(torch.max(x, inputs - self.epsilon), inputs + self.epsilon)
            x = torch.clamp(x, 0.0, 1.0)
        return  x, self.basic_net(x)
